### ISSUE TITLE
Big performance hit on operations mixing Float32 and Float64

### ISSUE THREAD

**Issue by ssfrr** [JuliaIntervals/IntervalArithmetic.jl#330]
I was just debugging a performance problem when I realized that my real input is made up of `Float32`s whereas my previous simulated testing had been with `Float64`s. This is on the `nonrecursive_powers` branch.

```julia
using BenchmarkTools
using IntervalArithmetic

setrounding(Interval, :none)
display(@benchmark $(Interval{Float64}(-1.5, 1.5)) + $(0.4f0))
display(@benchmark $(Interval{Float64}(-1.5, 1.5)) + $(0.4))
```

Which gives:

```
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     178.146 ns (0.00% GC)
  median time:      179.921 ns (0.00% GC)
  mean time:        186.607 ns (0.00% GC)
  maximum time:     536.108 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     705
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     0.021 ns (0.00% GC)
  median time:      0.024 ns (0.00% GC)
  mean time:        0.024 ns (0.00% GC)
  maximum time:     0.041 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
(v1.1) pkg> 
```

**Comment by dpsanders** [comment]
What about if the intervals are Float32 too?
Or just convert all your values to Float64? 

We never did extensive (=any) testing with Float32 so there are probably still lots of places where Float64 is implicitly assumed, even though I cleaned some of that up at some point.

**Comment by dpsanders** [comment]
Presumably this is not just due to Julia's own conversion from Float32 to Float64? That is presumably very fast.

**Comment by ssfrr** [comment]
It's still slow when everything is a `Float32`.

Looks like promoting/adding a `Float64` and `Float32` is on the order of 0.02ns, so that doesn't seem to be the issue.

It's definitely not a major problem for me - I can just convert my data to `Float64` first. Just logging a little speedbump I hit. :)

**Comment by dpsanders** [comment]
The problem is here:
```jl
julia> @btime convert(Interval{Float64}, 0.4f0)
  147.391 ns (0 allocations: 0 bytes)
[0.399999, 0.400001]
```

This has come up before. Thanks for reporting!

**Comment by dpsanders** [comment]
Basically: does `0.4f0` mean "the Float32 which we represent by `0.4f0`", or do you really mean "the true real number `0.4`"?

We try to be clever by interpreting it as the latter, which is where the performance hit is coming from, since then we need to do clever stuff.

I am increasingly thinking that we should stop doing this so magically.

**Comment by ssfrr** [comment]
Yeah, IMO the magic adds more ambiguity. Maybe a string macro would be a better way to do the more magic versions, something like `real"0.4"`? (I don't really know what expectations people who use this sort of thing would have)

**Comment by dpsanders** [comment]
Yep, in fact we already have two ways of doing that:
```
julia> @format full
Display parameters:
- format: full
- decorations: false
- significant figures: 6

julia> I"0.4"
Interval(0.39999999999999997, 0.4)

julia> @interval 0.4
Interval(0.39999999999999997, 0.4)
```

(The string macro is a capital `i` for `Interval`.)

**Comment by dpsanders** [comment]
But I do like being able to do `0.3..0.4` and getting an interval containing the *real* numbers `3//10` and `4//10`.

**Comment by lbenet** [comment]
Coming late to the party...

> But I do like being able to do 0.3..0.4 and getting an interval containing the real numbers 3//10 and 4//10.

I agree that the macro (*and* the `..` notation) should be used for assuring that the *real number* is contained in the returned interval, which involves certain loss of performance. Direct use of `Interval` or `interval` should be the responsibility of the user, acknowledging that it is faster because avoids all the rounding machinery

**Comment by dpsanders** [comment]
The question is whether the conversion of `0.4f0` to `Interval{Float64}` should round or not. Now I am thinking that it should not.

**Comment by lbenet** [comment]
We have `IntervalArithmetic.atomic`, which I think was design precisely for this, and does round with good performance.

```julia
julia> using IntervalArithmetic

julia> @format full;

julia> IntervalArithmetic.atomic(Interval{Float64}, 0.4)
Interval(0.39999999999999997, 0.4)

julia> IntervalArithmetic.atomic(Interval{Float32}, 0.4f0)
Interval(0.39999998, 0.4)

julia> convert(Interval{Float64}, 0.4)
Interval(0.4, 0.4)

julia> convert(Interval{Float32}, 0.4f0)
Interval(0.4, 0.4)

julia> 0.4 .. 0.4
Interval(0.39999999999999997, 0.4)

julia> 0.4f0 .. 0.4f0
Interval(0.39999998, 0.4)
```

**Comment by dpsanders** [comment]
Sorry I'm just getting back to this. `atomic` is slow:
```jl
julia> @btime IntervalArithmetic.atomic(Interval{Float64}, 0.4)
  152.968 ns (0 allocations: 0 bytes)
[0.399999, 0.400001]

julia> @btime IntervalArithmetic.atomic(Interval{Float64}, 0.4f0)
  156.447 ns (0 allocations: 0 bytes)
[0.399999, 0.400001]


julia> @btime $(Ref(interval(0.4, 0.4)))[]
  1.421 ns (0 allocations: 0 bytes)
[0.4, 0.400001]
```
Note that benchmark times of 0.0 ns are wrong: no calculation is actually being done.
You need to use this annoying `Ref` business to get the correct time.

`convert` should definitely not do anything "clever"; it should just construct an interval.
So it should usually bypass `atomic` I think.

**Comment by Kolaru** [comment]
Every `Float32` should be exactly representible as a `Float64` so there is no reason for any rounding there.

The other direction (`Float64` to `Float32`, or any conversion with loss of the precision) we need to take care of rounding, but that kind of conversion should never happen automatically. We could go as far as returning an error when such conversion is attempted (forcing the use of the constructors), but a more reasonnable approach may be to use `atomic` only in that case (and all other cases where precision is lost in the conversion).
